File: sdk/php/src/Providers/TempyEmail.php
<?php

declare(strict_types=1);

namespace ChanhanzhanX\TempMail\Providers;

use ChanhanzhanX\TempMail\Email;
use ChanhanzhanX\TempMail\EmailInfo;
use ChanhanzhanX\TempMail\HttpClient;
use ChanhanzhanX\TempMail\Normalize;

/**
 * Tempy Email 渠道实现 — https://tempy.email
 *
 * POST /api/v1/mailbox 创建邮箱；GET /api/v1/mailbox/{email}/messages 收信。
 */
final class TempyEmail
{
    private const CHANNEL = 'tempy-email';
    private const API_BASE = 'https://tempy.email/api/v1';

    /** @var array<string,string> */
    private const HEADERS = [
        'Accept' => 'application/json',
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36',
    ];

    /**
     * @param array<mixed> $raw
     * @return array<mixed>
     */
    private static function flatten(array $raw, string $recipient): array
    {
        return [
            'id' => $raw['id'] ?? $raw['messageId'] ?? $raw['message_id'] ?? $raw['mail_id'] ?? '',
            'from' => $raw['from'] ?? $raw['sender'] ?? $raw['from_addr'] ?? $raw['from_address'] ?? '',
            'to' => $raw['to'] ?? $recipient,
            'subject' => $raw['subject'] ?? $raw['mail_title'] ?? '',
            'text' => $raw['text'] ?? $raw['body_text'] ?? $raw['text_body'] ?? $raw['body'] ?? '',
            'html' => $raw['html'] ?? $raw['body_html'] ?? $raw['html_body'] ?? '',
            'date' => $raw['date'] ?? $raw['received_at'] ?? $raw['created_at'] ?? '',
            'is_read' => $raw['is_read'] ?? $raw['isRead'] ?? $raw['seen'] ?? false,
            'attachments' => $raw['attachments'] ?? [],
        ];
    }

    public static function generate(?string $domain = null): EmailInfo
    {
        $body = [];
        $wanted = trim((string) ($domain ?? ''));
        if ($wanted !== '') {
            $body['domain'] = $wanted;
        }
        $resp = HttpClient::post(
            self::API_BASE . '/mailbox',
            self::HEADERS + ['Content-Type' => 'application/json'],
            json: $body ?: new \stdClass(),
        );
        $data = HttpClient::json($resp);
        $email = trim((string) ($data['email'] ?? ''));
        if ($email === '') {
            throw new \RuntimeException('tempy-email: 无效的创建响应');
        }
        return new EmailInfo(self::CHANNEL, $email);
    }

    /**
     * @return Email[]
     */
    public static function getEmails(string $email, ?string $token = null): array
    {
        $addr = trim($email);
        if ($addr === '') {
            throw new \InvalidArgumentException('tempy-email: 邮箱地址为空');
        }
        $resp = HttpClient::get(self::API_BASE . '/mailbox/' . rawurlencode($addr) . '/messages', self::HEADERS);
        $data = HttpClient::json($resp);
        $rows = $data['messages'] ?? $data;
        if (!is_array($rows)) {
            return [];
        }
        $out = [];
        foreach ($rows as $item) {
            if (is_array($item)) {
                $out[] = Normalize::email(self::flatten($item, $addr), $addr);
            }
        }
        return $out;
    }
}

File: sdk/csharp/src/Providers/TempyEmail.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json.Nodes;

namespace XxxXTeam.TempMail.Providers;

/// <summary>
/// Tempy Email 渠道（tempy.email/api/v1）。
/// 创建：POST /mailbox（可选 domain）；收信：GET /mailbox/{email}/messages，多候选字段扁平化。
/// </summary>
public static class TempyEmail
{
    private const string ApiBase = "https://tempy.email/api/v1";

    private static Dictionary<string, string> Headers() => new()
    {
        ["Accept"] = "application/json",
        ["User-Agent"] = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36",
    };

    private static string Pick(JsonObject raw, params string[] keys)
    {
        foreach (var k in keys)
        {
            var v = Json.Str(raw, k);
            if (!string.IsNullOrEmpty(v)) return v;
        }
        return "";
    }

    private static Dictionary<string, object?> Flatten(JsonObject raw, string recipient)
    {
        var to = Json.Str(raw, "to");
        return new Dictionary<string, object?>
        {
            ["id"] = Pick(raw, "id", "messageId", "message_id", "mail_id"),
            ["from"] = Pick(raw, "from", "sender", "from_addr", "from_address"),
            ["to"] = string.IsNullOrEmpty(to) ? recipient : to,
            ["subject"] = Pick(raw, "subject", "mail_title"),
            ["text"] = Pick(raw, "text", "body_text", "text_body", "body"),
            ["html"] = Pick(raw, "html", "body_html", "html_body"),
            ["date"] = Pick(raw, "date", "received_at", "created_at"),
            ["is_read"] = (raw["is_read"] as JsonValue)?.GetValue<bool>()
                ?? (raw["isRead"] as JsonValue)?.GetValue<bool>()
                ?? (raw["seen"] as JsonValue)?.GetValue<bool>() ?? false,
            ["attachments"] = Json.ToRaw(raw["attachments"]),
        };
    }

    /// <summary>创建邮箱</summary>
    public static EmailInfo Generate(string? domain)
    {
        var body = new Dictionary<string, object?>();
        var wanted = (domain ?? "").Trim();
        if (wanted.Length > 0) body["domain"] = wanted;
        var headers = Headers();
        var resp = Http.Post($"{ApiBase}/mailbox", Json.Serialize(body), "application/json", headers);
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body) as JsonObject ?? throw new Exception("tempy-email: invalid create response");
        var email = Json.Str(data, "email").Trim();
        if (email.Length == 0) throw new Exception("tempy-email: invalid create response");
        return new EmailInfo("tempy-email", email);
    }

    /// <summary>获取邮件列表</summary>
    public static List<Email> GetEmails(string email)
    {
        var address = (email ?? "").Trim();
        if (address.Length == 0) throw new Exception("tempy-email: empty email");
        var resp = Http.Get($"{ApiBase}/mailbox/{WebUtility.UrlEncode(address)}/messages", Headers());
        resp.EnsureSuccess();
        var data = Json.Parse(resp.Body);
        var rows = (data as JsonObject)?["messages"] as JsonArray ?? data as JsonArray;
        var result = new List<Email>();
        if (rows is null) return result;
        foreach (var item in rows)
            if (item is JsonObject ro) result.Add(Normalize.NormalizeEmail(Flatten(ro, address), address));
        return result;
    }
}
